File: backend/routers/memory.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from dependencies import get_db, verify_api_key
from services.memory_service import MemoryService
from schemas.memory import MemoryCreate

router = APIRouter(tags=["Memory"])


@router.post("/memory", dependencies=[Depends(verify_api_key)])
async def create_memory(body: MemoryCreate, db: AsyncSession = Depends(get_db)):
    svc = MemoryService(db)
    return await svc.create(body.type, body.content, body.session_id, body.tags)


@router.get("/memory/search", dependencies=[Depends(verify_api_key)])
async def search_memory(query: str = Query(...), type: str | None = None, limit: int = 10, db: AsyncSession = Depends(get_db)):
    svc = MemoryService(db)
    return await svc.search(query, type, limit)


@router.get("/memory/{memory_id}", dependencies=[Depends(verify_api_key)])
async def get_memory(memory_id: str, db: AsyncSession = Depends(get_db)):
    svc = MemoryService(db)
    result = await svc.get(memory_id)
    if not result:
        raise HTTPException(status_code=404, detail="Memory not found")
    return result


@router.delete("/memory/{memory_id}", dependencies=[Depends(verify_api_key)])
async def delete_memory(memory_id: str, db: AsyncSession = Depends(get_db)):
    svc = MemoryService(db)
    ok = await svc.delete(memory_id)
    if not ok:
        raise HTTPException(status_code=404, detail="Memory not found")
    return {"success": True, "message": "Memory deleted"}

File: frontend/src/api/memory.ts
import { api } from './client';

export async function createMemory(type: string, content: string, tags: string[] = [], sessionId?: string) {
  return api.post<{ memory_id: string; type: string; content: string; created_at: string }>('/memory', { type, content, tags, session_id: sessionId });
}

export async function searchMemory(query: string, type?: string, limit = 10) {
  const params = new URLSearchParams({ query, limit: String(limit) });
  if (type) params.append('type', type);
  return api.get<{ results: Array<{ memory_id: string; type: string; content: string; similarity: number }> }>(`/memory/search?${params}`);
}

export async function getMemory(memoryId: string) {
  return api.get(`/memory/${memoryId}`);
}

export async function deleteMemory(memoryId: string) {
  return api.delete(`/memory/${memoryId}`);
}
